DEV: Analysis refactoring

fernandezc opened the pull request:
This PR is a proposal to rewrite the MCRALS class in a more modular way and taking advantage
of the potentialities of the traitlets library that we use everywhere else in the SpectroChemPy core.

The MCRALS class was already a subclass of traitlets.HasTraits but did not take advantage of it.
However, as we can see in this rewriting proposal, we can use at least two features.

- automatic validation of special class attributes (Traits) according to their predefined type: indeed, if we take
  the example of the 'closureTarget' parameter, which can be either 'default', a string, but
  not any other, or a list, the Traits will be defined like this

  ```python
  closureTarget = Union(
          (Enum(["default"]), List()),
          default_value="default")
  ```
  This way it will not be possible to use another string other than the one provided, otherwise an error message will be displayed.

  This validation can be completed by a custom validator (decorator `validate`) that will be triggered automatically after the first automatic.

  ```Python
  @validate("closureTarget")
  def _validate_closureTarget(self, proposal):
      closureTarget = proposal.value
      ny = self.X.shape[0]
      if closureTarget == "default":
          closureTarget = np.ones(ny)
      elif len(closureTarget) != ny:
          raise ValueError(
              f"The data contain only {ny} observations, please check "
              f"the 'closureTarget' configuration (value:{closureTarget})"
          )
      return closureTarget
  ```
  This can be done for all the parameters that need to be validated before starting the execution of the program.

- The second thing we can use is the `observe` decorator which allows to automatically trigger an action,
  when a Traits attribute of the class changes. We can use it to initialize the C and St data every time we
  change a value, without having to write all the machinery.
  In MCRALS, I only used it once, but there are probably other possibilities.

You can stay with an MCRALS class derived from HasTraits, but in my opinion there is even better: the `Configurable`
class which is also derived from HasTraits. The advantage is that some Traits are defined as configuration parameters
(with a particular tag).

```python
maxdiv = Integer( 5, help="Maximum number of successive non-converging iterations." ).tag(config=True)
```
These configuration traits have default values that can be defined in a separate file and loaded each time the
MCRALS class is created. For this object it may not be necessary but it allows for example to change some defaults permanently.
(I haven't finished this yet because it doesn't seem to me the most urgent).

This said Arnaud you can see already in this PR if you find it useful. I tried to comment the program as best as
I could so that you know the reasons for the changes. The `test_mcr_als.py` file has also been modified a bit according
to the new features added.

The usage files you wrote before are ALMOST compatible as you can see in the example `userguide/analysis/mcr_als.py`.

The main differences are the following:

* when you wrote:
  ```python
  mcr = MCRALS(DATA,ST0,maxit=30)
  ```
  The subsequent calculation was executed automatically.

  Now you have to do this:
    ```python
  mcr = MCRALS(X0,ST0,maxit=30)
  mcr.run()   # <<<<<<<<<<<<<<<<<< Addition to execute the calculation
  ```
  The advantage is the following with a new worflow:

  ```python
  mcr = MCRALS()

  mcr.X = DATA
  mcr.C = C0
  mcr.run()    # take the default `tol=0.1`

  mcr.tol=0.001   # new value
  mcr.run()       # just run it

  ```

  That's it for now. I'll let you look at it and say what you think.

  One REAL advantage I haven't mentioned yet is that the automatic validation allows you to spot errors
  that you would hardly see. There are some minor ones in your version.

  If you think it's OK, we can make a merge, but we still have to check the documentation and the examples,
  and I'd probably like to add the reading/writing of config files and a reset function to restore the default
  values between different runs, because it's not easy.

  If it suits you we can also write in an organized way the other classes like IRIS etc....

fernandezc commented:
To test it, you just need to checkout my branch on the fernandezc/spectrochempy  repository. You can even push directly to it.

atravert commented:
Thank you very much Christian,

I will look at this !   I wanted to refactor/rewrite the dimentionality 
reduction methods currently implemented in scpy (SVD, PCA, MCRALS, NNMF, 
SIMPLISMA, EFA) to make them more consistent with each other, and maybe 
replacing some of them by wrappers of their equivalent in scikit-learn 
(SVD, PCA, NMF), so it could be the occasion to do all in once ?



Le 09/02/2023 à 23:58, Christian Fernandez a écrit :
>
> This PR is a proposal to rewrite the MCRALS class in a more modular 
> way and taking advantage
> of the potentialities of the traitlets library that we use everywhere 
> else in the SpectroChemPy core.
>
> The MCRALS class was already a subclass of traitlets.HasTraits but did 
> not take advantage of it.
> However, as we can see in this rewriting proposal, we can use at least 
> two features.
>
>  *
>
>     automatic validation of special class attributes (Traits)
>     according to their predefined type: indeed, if we take
>     the example of the 'closureTarget' parameter, which can be either
>     'default', a string, but
>     not any other, or a list, the Traits will be defined like this
>
>     closureTarget  =  Union(
>              (Enum(["default"]),List()),
>              default_value="default")
>
>     This way it will not be possible to use another string other than
>     the one provided, otherwise an error message will be displayed.
>
>     This validation can be completed by a custom validator (decorator
>     |validate|) that will be triggered automatically after the first
>     automatic.
>
>     @validate("closureTarget")
>     def  _validate_closureTarget(self,proposal):
>          closureTarget  =  proposal.value
>          ny  =  self.X.shape[0]
>          if  closureTarget  ==  "default":
>              closureTarget  =  np.ones(ny)
>          elif  len(closureTarget)!=  ny:
>              raise  ValueError(
>                  f"The data contain only {ny} observations, please check "
>                  f"the 'closureTarget' configuration (value:{closureTarget})"
>              )
>          return  closureTarget
>
>     This can be done for all the parameters that need to be validated
>     before starting the execution of the program.
>
>  *
>
>     The second thing we can use is the |observe| decorator which
>     allows to automatically trigger an action,
>     when a Traits attribute of the class changes. We can use it to
>     initialize the C and St data every time we
>     change a value, without having to write all the machinery.
>     In MCRALS, I only used it once, but there are probably other
>     possibilities.
>
> You can stay with an MCRALS class derived from HasTraits, but in my 
> opinion there is even better: the |Configurable|
> class which is also derived from HasTraits. The advantage is that some 
> Traits are defined as configuration parameters
> (with a particular tag).
>
> maxdiv  =  Integer(5,help="Maximum number of successive non-converging iterations."  ).tag(config=True)
>
> These configuration traits have default values that can be defined in 
> a separate file and loaded each time the
> MCRALS class is created. For this object it may not be necessary but 
> it allows for example to change some defaults permanently.
> (I haven't finished this yet because it doesn't seem to me the most 
> urgent).
>
> This said Arnaud you can see already in this PR if you find it useful. 
> I tried to comment the program as best as
> I could so that you know the reasons for the changes. The 
> |test_mcr_als.py| file has also been modified a bit according
> to the new features added.
>
> The usage files you wrote before are ALMOST compatible as you can see 
> in the example |userguide/analysis/mcr_als.py|.
>
> The main differences are the following:
>
>  *
>
>     when you wrote:
>
>     mcr  =  MCRALS(DATA,ST0,maxit=30)
>
>     The subsequent calculation was executed automatically.
>
>     Now you have to do this:
>
>     mcr  =  MCRALS(X0,ST0,maxit=30)
>     mcr.run()# <<<<<<<<<<<<<<<<<< Addition to execute the calculation
>
>     The advantage is the following with a new worflow:
>
>     mcr  =  MCRALS()
>
>     mcr.X  =  DATA
>     mcr.C  =  C0
>     mcr.run()# take the default `tol=0.1`
>
>     mcr.tol=0.001    # new value
>     mcr.run()# just run it
>
>     That's it for now. I'll let you look at it and say what you think.
>
>     One REAL advantage I haven't mentioned yet is that the automatic
>     validation allows you to spot errors
>     that you would hardly see. There are some minor ones in your version.
>
>     If you think it's OK, we can make a merge, but we still have to
>     check the documentation and the examples,
>     and I'd probably like to add the reading/writing of config files
>     and a reset function to restore the default
>     values between different runs, because it's not easy.
>
>     If it suits you we can also write in an organized way the other
>     classes like IRIS etc....
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/spectrochempy/spectrochempy/pull/576 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148668344%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=JAphvBb%2BGmpo9uUaMn7nRbEQK560DZ897XsQqdkcYi0%3D&reserved=0>
>
>
>         Commit Summary
>
>   * 0191988
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Fcommits%2F0191988e40fae99d03818690edbcede0bbb9aef5&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148668344%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=OaqnLiUjV5aEGDSGKkVw9oJd14mh3Giuu25m8wgpewM%3D&reserved=0>
>     DEV: MCRALS refactoring proposal
>
>
>         File Changes
>
> (8 files 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=fHZmz5fLVmmc8VQZK9OejPyHOoLR8A14hz5t3HjqWe0%3D&reserved=0>)
>
>   * *M* docs/gettingstarted/examples/nddataset/plot_create_dataset.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-a75f6e1fca18217078ce456d1e70f5bd89c77cdfef85f4b6020fc11aace911b5&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=u2tXDd0ZuYsbgfYlKiDoh8cNLVMphL33pOTUqrkm8ec%3D&reserved=0>
>     (1)
>   * *M* docs/userguide/analysis/mcr_als.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-928bca4f69eba2d1988d5bd5a5d9b0d5b7ea8499958714d0a91659a75d0d11a3&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=43lGG02z6VyYA8l3zzxUfX8DXMI2oO4dMS%2Bzh4HT5r4%3D&reserved=0>
>     (54)
>   * *M* docs/userguide/introduction/introduction.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-7c42e431de952614319c3c12060e5753195aec8226bd327ba728690bcd0fe9ac&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=uGqhqiGJjG1uH76S8B6fAMKr1K9WL%2BrDmRWQaT5MieU%3D&reserved=0>
>     (4)
>   * *M* setup.cfg
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=InNo9arbN0MLpMkyaCeE5Q%2Bs%2FZtD9l4%2BXh%2BS3oL%2Fpg0%3D&reserved=0>
>     (8)
>   * *M* spectrochempy/analysis/mcrals.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-954597280c04e359fc62d00c4a3e68d7d7a6344d83f4f19e19fa4da770dca72d&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=jJ9Jrz6iGbu6hMK%2FiY%2BHUhVYUR%2Bvnif%2Fovm05qBO6V4%3D&reserved=0>
>     (1088)
>   * *M* spectrochempy/application.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-a9edebf1171287850405aa0ec1e62861551573925b16f1b3763e3e0e22f079d5&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=g%2FSnJbOjB32tDuwo%2BzLY6NGfG7IW5ZgCIRRIkk0zrHA%3D&reserved=0>
>     (17)
>   * *M* spectrochempy/utils/traits.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-5a5740bd67d25f5df3fdbb81b816b10d8f90ad25225b984fb0d9d7a35ff78f17&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=RJeemSQUagxGvl0f0TCIKFrQYolvKGacmfcWJJ38y8M%3D&reserved=0>
>     (6)
>   * *M* tests/test_analysis/test_mcrals.py
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576%2Ffiles%23diff-9a4281f49c6ec4f9f7c7bcb9a3f798f5f7cf41aa8086b6d626d5f77aff0ecccd&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=xCXn%2BjP2hOxMRzm0tV%2F8S8tCP7jdEJCRgRGBg2Z61ac%3D&reserved=0>
>     (194)
>
>
>         Patch Links:
>
>   * https://github.com/spectrochempy/spectrochempy/pull/576.patch
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576.patch&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=2QF5lDwPrMf2O0Uc6W0bpl7GXUML3ej5IRezqTS1eZg%3D&reserved=0>
>
>   * https://github.com/spectrochempy/spectrochempy/pull/576.diff
>     <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576.diff&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=1XD4D8s7lfwekbgiJXZE0as%2B%2FfLKEfA7LwqdX8QNhR8%3D&reserved=0>
>
>
> —
> Reply to this email directly, view it on GitHub 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F576&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=KBMMrdAsa9MxFloe2URlR%2FPPxWbCaX%2F3PI8DbeOTLWI%3D&reserved=0>, 
> or unsubscribe 
> <https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCVZ7PHYSAQ36BXETO3WWVZBNANCNFSM6AAAAAAUXECDP4&data=05%7C01%7Carnaud.travert%40ensicaen.fr%7C7d7fed365f474da514a908db0af12a24%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C638115803148824617%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=MRXdw%2FJaAwjDxUhgdeqz%2BfemWIkikLIsUIc%2BfEgupU8%3D&reserved=0>.
> You are receiving this because your review was requested.Message ID: 
> ***@***.***>
>

fernandezc commented:
👍 Yes, this is something I was thinking of doing from the beginning regarding Scikit-learn, but I thought you didn't want to have any dependency on this library. To do that, it seems to me quite simple, the wrapper can be written like I do for NumPy functions with a generic decorator that allows to take into account the structure of the datasets with the units etc... I'm actually in the process of refactoring the core to make it more readable (especially since NumPy offers tools like __array_function__ and __array_ufunc__ that are very useful. )
I've almost rewritten the whole NDMath class, where a lot of tricky things became obsolete with this process. 
You can do equivalent things with other libraries like scikit-learn or others. 
AND it is perfectly compatible with the techniques used in this PR.

If you want I can write you a little template to start with.(?)

fernandezc commented:
Some changes:
 The run method has been replaced by fit (especially if we want to homogenize some functions with scikit-learn's one later on. 
The configuration is stored in a json file which is reused if we put the keyword warm-start=True in the class constructor, if warm_start is False (default) then the configuration is reset to default.  (needs to add few tests)

fernandezc commented:
Output C and St were incorrect. Fixed now.

fernandezc commented:
Well, this is probably a bit anecdotal for a class like MCRALS, but now the configuration is saved in the `config` directory in `~/.spectrochempy`.
The default values file is `MCRALS.py`  file (created after at least one run of SpectroChempy), which lists all configurable values. If the user wants to change one of them without touching the code, he just has to change the value in the MCRALS.py file on the corresponding line and uncomment the line (This is True also for the other configuration file like `PlotPreferences.py`.

for example, if we want to change the value of `maxit`

```python
## Maximum number of ALS iteration
# Default: 50
# c.MCRALS.maxit = 50
```

becomes 

```python
## Maximum number of ALS iteration
# Default: 50
c.MCRALS.maxit = 30  # <--- look at the uncommented line!
```

There is also a json file in this config directory, created after the first run, with the same name, which reports the on-the-fly values at each variable change. You can eventually use it to recall a configuration between files (I'm not sure if it's useful, but the possibility is there - just set `warm_start` to True in the constructor.

fernandezc commented:
In the last commit, I extracted all the parts that can be common in the development of the methods, like the log system, the helper functions, the plotmerit, ... So to homogenize with PCA for example, it's enough to write just the typical functions like fit, reconstruct, .... the rest happens in the super class.

fernandezc commented:
The structure set up for the homogenization of the analysis models is the following.

1) each decomposition model must derive (as it is already the case for PCA and MCRALS) from the DecompositionAnalysisConfigurable class, which itself derives from the AnalysisConfigurable class. Classes that do not fall within the scope of decomposition models can be derived directly from AnalysisConfigurable.

2) the AnalysisConfigurable class has for public method, the fit method which as in sckit-learn launches the basic process. This function must not be redefined in the subclasses (see example PCA and MCRALS) but only the private method _fit which is specific to each model. The fit method always takes an NDDataset (X) as input but the _fit method which is called internally only works on the data (np.ndarray): This keeps the performances as good as possible by avoiding the interpretation of the metadata at each step of the calculation.
The fit method can take one (X) as in PCA or two (Y) parameters as in MCRALS. 
Y is either an NDDataset or a tuple of NDDatasets.
Note that unlike sklearn, we handle masked data in fit. Masked rows or columns are removed before _fit is called, and then reintroduced when the calculation data is returned later.
The output of fit is stored internally in the private variable _outfit as np.ndarray.
At this point in the process, the data is raw (no metadata like coordinates or units).

3) In the derived class DecompositionAnalysisConfigurable
there are public methods designed for decomposition models: transform (reduce) and inverse_transform (reconstruct) and the components property (with the associated get_components method). As with fit, we do not touch these methods in the subclasses. We just define the private methods, _transform and _inverse transform, which works with ndarray not dataset. Similarly, _get_components is used to output the number of components.
The data to be returned to the user is wrapped by the utility decorator _wrap_ndarray_output_to_nddataset.
In this way, NDDatasets are returned without any intervention other than the possible addition of metadata.
Other features of the program changes can be found in the examples or tests

fernandezc commented:
@atravert, I think you can now review the changes. Apart from the problem with the doc I'm not going to change anything for the moment. The terminology changes significantly from the existing one, but I think it's better this way. To be discussed of course.

fernandezc commented:
@atravert, new since my last post:
- SIMPLISMA adapted (I did not test interactivity)
- Score plots with labels (as requested in a discussion)
- Some corrections and improvements of the code.
Use this version, I think, to try and review.

fernandezc commented:
@atravert Since my last post, there have been several corrections, and especially the addition of SVD and EFA recast in DecompositionAnalysis style. This is of course a first version which can evolved/corrected/improved  if I have some feedback from you.
In the meantime, I'll continue, because I think it's quite interesting to have a kind of common base for all methods, even if it will change the habits of some of the current users.

fernandezc commented:
@travert I updated the PR with the IRIS refactoring to conform to the other models. Now all the analysis models look similar:

* Initialize the model (with some configuration parameters): for example iris = IRIS(reg_par=None).
* In the case of IRIS, the kernel is computed independently: using K = IrisKernel(X, "Langmuir", etc...)
* then the model is fitted: iris.fit(X, K)

IRIS can now work with a mask on the X.x dimensions (I didn't really check the masking of the X rows). you can look at the tests for more information.

While doing this, I found some errors in the main code that I will publish as soon as possible in the main branch (for example, the multiple coordinate slicing was not working well).

Now, as far as redesigning the analysis, I think the only thing left to do is to adapt the nonlinear fit. Previously I had put FastICA but I have a problem with the coordinates so for the moment I removed it. Anyway it's maybe not very useful to analyze the spectra. 

And something important, we need to complete the documentation of the doc with more examples and pages in the user guide.

fernandezc commented:
The Fit method has been renamed Optimize and redesigned along the same lines as the other decomposition methods. This is certainly a work in progress as there is room for many improvements. We can certainly call other libraries like iminuit or lmfit, but I think that we already call a little too many external libraries which is already a lot of dependencies (so at first I think I'll try to avoid it). Anyway in this rewrite I tried to answer the comments of the discussion #393

fernandezc commented:
@atravert Arnaud I'm currently working on the new analysis model docs. If you want to see the results without compiling it yourself, you can see the current update after each commits here: https://fernandezc.github.io/spectrochempy/latest/userguide/reference/index.html#analysis

for exemple (for EFA doc)
https://fernandezc.github.io/spectrochempy/latest/userguide/reference/generated/spectrochempy.EFA.html#spectrochempy.EFA

More work is needed, yet!

fernandezc commented:
The analysis API load time is severely degraded! 11 seconds on my Mac. I'm going to investigate the cause.

fernandezc commented:
The error in plotmerit was the use of self._X (with mask removed) instead of self.X (with masks restored).
self.inverse_transform also restore mask. 

Doing this way we get a much better display:

![image](https://user-images.githubusercontent.com/2428474/231445907-e708ef0a-6a26-4bfb-ab29-98210c0941d2.png)

instead of: 
![image](https://user-images.githubusercontent.com/2428474/231445807-c70719b9-f7e7-4271-b328-e51c474210d4.png)

fernandezc commented:
@atravert I think that even if there are certainly still things to refine the PR become much too long and until we have done the merge it is difficult to intervene on the master branch. Have you finished testing and can we do the merge? Then we can release a version 0.6

atravert commented:
Yes !, thank you again for the huge work. I am working on adding PLS: a 
very good exercise for me, and very useful and used method :-)

Le 13/04/2023 à 08:00, Christian Fernandez a écrit :
>
> @atravert <https://github.com/atravert> I think that even if there are 
> certainly still things to refine the PR become much too long and until 
> we have done the merge it is difficult to intervene on the master 
> branch. Have you finished testing and can we do the merge? Then we can 
> release a version 0.6
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/spectrochempy/spectrochempy/pull/576#issuecomment-1506391038>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACNBNCX6YBR44XWGFOVDXTTXA6I6NANCNFSM6AAAAAAUXECDP4>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

fernandezc commented:
replaced by PR #611
